### PR TITLE
Node e2e Makefile support for running remote tests against kubernetes…

### LINKED ISSUE #26665: Make node e2e test images publicly accessible
Node e2e test images should be publicly accessible.  The right solution may be to create a dedicated project for this and to have it be the canonical source.

### PR DESCRIPTION
**Pull request by pwittrock**
Fixes #26665

[Analytics]()

…-node-e2e-images.

### CODE REVIEW

**Comment by pwittrock** [comment]
Only documentation is remaining

**Comment by pwittrock** [comment]
Added the documentation updates.

File: docs/devel/e2e-node-tests.md [hunk 1]
```diff
@@ -34,147 +34,180 @@ Documentation for other releases can be found at
...
+  - kubelet
+  - kube-apiserver
+  - e2e_node.test (this binary contains the actual tests to be run)
+- Unzip the archive to a directory under /tmp/gcloud*
+- Run the tests using the `ginkgo` command
+  - Starts etcd, kube-apiserver, kubelet
+  - The ginkgo command is used because this supports more features than running the test binary directly
+- Output the remote test results to STDOUT
+- `scp` the log files back to the local host under /tmp/_artifacts/e2e-node-containervm-v20160321-image
+- Delete the created GCE instance
```
Comment (vishh) [inline docs/devel/e2e-node-tests.md]: Didn't we want to keep the instance around by default?

File: docs/devel/e2e-node-tests.md [hunk 2]
```diff
@@ -34,147 +34,180 @@ Documentation for other releases can be found at
...
+This is useful if you want to inspect or debug the host after the tests complete, or if you want to run
+additional tests without recreating a new instance each time the tests are run.
+**Note: When not-deleting an instance, subsequent tests run using the same image will *reuse the existing host*,
+even if they would normally provision a new instance.**
 
-    * Make sure this works - e.g. `ssh <hostname> -- sudo echo "ok"`
-    * If ssh flags are required (e.g. `-i`), they can be used and passed to the
-tests with `--ssh-options`
+```sh
+make test_e2e_node REMOTE=true DELETE_INSTANCES=false
```
Comment (vishh) [inline docs/devel/e2e-node-tests.md]: Can we make this the default?

File: Makefile [hunk 3]
```diff
@@ -99,10 +99,9 @@ test_e2e:
 #   make test_e2e_node FOCUS=kubelet SKIP=container
 # Build and run tests.
 test_e2e_node:
-	hack/e2e-node-test.sh FOCUS=$(FOCUS) SKIP=$(SKIP)
+	hack/e2e-node-test.sh
```
Comment (vishh) [inline Makefile]: nit: All the options that can be passed to this makerule should be documented above. That way once we add `make help`, it is self-sufficient.

File: docs/devel/e2e-node-tests.md [hunk 4]
```diff
@@ -34,147 +34,180 @@ Documentation for other releases can be found at
...
-| ubuntu          | trusty         | 1.9            |              | gce            |
-| ubuntu          | trusty         | 1.10           |              | gce            |
+## Run tests continually until they fail
 
+This is useful if you are trying to debug a flaky test failure.  This will cause ginkgo to continually
+run the tests until they fail.  **Note: this will only perform test setup once (e.g. creating the instance) and is
+less useful for catching flakes related creating the instance from an image.**
 
+```sh
+make test_e2e_node REMOTE=true RUN_UNTIL_FAILURE=true
```
Comment (vishh) [inline docs/devel/e2e-node-tests.md]: :100: for adding this.

File: docs/devel/e2e-node-tests.md (hunk 1 above)
Comment (pwittrock) [inline docs/devel/e2e-node-tests.md]: Done

File: docs/devel/e2e-node-tests.md (hunk 2 above)
Comment (pwittrock) [inline docs/devel/e2e-node-tests.md]: Done

File: docs/devel/e2e-node-tests.md [hunk 5]
```diff
@@ -34,147 +34,180 @@ Documentation for other releases can be found at
...
-These steps should be captured in [setup_host.sh](../../test/e2e_node/environment/setup_host.sh)
+Why Run tests *Remotely*?  Tests will be run in a customized pristine environment.  Closely mimics what will be done as pre- and post- submit testing performed by the project.
 
-### Option 2: Copying an existing host image from another project
+Prerequisites:
+- [join the googlegroup](https://groups.google.com/forum/#!forum/kubernetes-dev)
+`kubernetes-dev@googlegroups.com`
+  - *This provides read access to the node test images.*
+- Setup a [Google Cloud Platform](https://cloud.google.com/) account and project with Google Compute Engine enabled
+- Install and setup the [gcloud skd](https://cloud.google.com/sdk/downloads)
```
Comment (mtaufen) [inline docs/devel/e2e-node-tests.md]: Typo: `gcloud skd` should be `gcloud sdk`

File: Makefile (hunk 3 above)
Comment (pwittrock) [inline Makefile]: Done

File: docs/devel/e2e-node-tests.md (hunk 2 above)
Comment (pwittrock) [inline docs/devel/e2e-node-tests.md]: Did you mean just the DELETE_INSTANCES=false part, or both that and REMOTE=true?

File: docs/devel/e2e-node-tests.md (hunk 4 above)
Comment (pwittrock) [inline docs/devel/e2e-node-tests.md]: woot.

File: docs/devel/e2e-node-tests.md (hunk 5 above)
Comment (pwittrock) [inline docs/devel/e2e-node-tests.md]: Done

File: test/e2e_node/runner/run_e2e.go [hunk 6]
```diff
@@ -233,8 +305,23 @@ func createInstance(image string) (string, error) {
 			err = fmt.Errorf("Instance %s not in state RUNNING, was %s.", name, instance.Status)
 			continue
 		}
+		externalIp := ""
+	Loop:
```
Comment (vishh) [inline test/e2e_node/runner/run_e2e.go]: nit: Can we move this to a function to avoid duplicating this logic?

**Comment by vishh** [comment]
Just a few more comments. Can't wait for this PR to get in :grinning:

File: hack/e2e-node-test.sh [hunk 7]
```diff
@@ -20,14 +20,110 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
...
+  if [ ! -d "${artifacts}" ]; then
+    echo "Creating artifacts directory at ${artifacts}"
+    mkdir -p ${artifacts}
+  fi
+  echo "Test artifacts will be written to ${artifacts}"
+
+  z="$(gcloud config list compute/zone 2>/dev/null | grep zone)"
+  zone_regex="zone\s*=\s*(.+)"
+  if [[ $z =~  $zone_regex ]]; then
+    zone="${BASH_REMATCH[1]}"
```
Comment (timstclair) [inline hack/e2e-node-test.sh]: prefer:
```
zone=$(gcloud config list compute/zone --format='value(compute.zone)' 2> /dev/null)
```

Same pattern applies below

File: test/e2e_node/e2e_build.go [hunk 8]
```diff
@@ -125,6 +125,14 @@ func getKubeletServerBin() string {
 	return bin
 }
 
+func getGinkgoBin() string {
+	bin, err := getK8sBin("ginkgo")
+	if err != nil {
+		panic(fmt.Sprintf("Could not locate ginkgo binary."))
```
Comment (timstclair) [inline test/e2e_node/e2e_build.go]: nit: prefer `log.Panicf`, `log.Fatalf` or `glog.Fatalf`

Comment (timstclair) [inline test/e2e_node/e2e_build.go]: Also: include err in the message

File: test/e2e_node/runner/run_e2e.go [hunk 9]
```diff
@@ -159,9 +174,62 @@ func main() {
...
+	for i := range instance.NetworkInterfaces {
+		ni := instance.NetworkInterfaces[i]
+		for j := range ni.AccessConfigs {
+			ac := ni.AccessConfigs[j]
+			if len(ac.NatIP) > 0 {
+				externalIp = ac.NatIP
+				break Loop
+			}
+		}
+	}
```
Comment (timstclair) [inline test/e2e_node/runner/run_e2e.go]: nit: I prefer helper functions for breaking out of nested loops

File: test/e2e_node/runner/run_e2e.go [hunk 10]
```diff
@@ -159,9 +174,62 @@ func main() {
...
+	if len(externalIp) > 0 {
+		e2e_node.AddHostnameIp(host, externalIp)
+	}
+
+	path, err := arc.getArchive()
+	if err != nil {
+		// Don't log fatal because we need to do any needed cleanup contained in "defer" statements
+		return &TestResult{
+			err: fmt.Errorf("Unable to create test archive %v.", err),
```
Comment (timstclair) [inline test/e2e_node/runner/run_e2e.go]: nit: error style starts with lowercase

**Comment by timstclair** [comment]
Handful of nits, but this is great!

**Comment by timstclair** [comment]
FYI - I ran into an issue testing against an existing host (created by a previous remote run with `DELETE_INSTANCES=false`). The error is that etcd can't run since the port is in use (I assume by etcd), but I verified that etcd is not running and the port is available prior to starting the test. Also surprising is that the suite fails 3 times, and then passes (why is it rerun at all?).

Log from results: 
[results.txt](https://github.com/kubernetes/kubernetes/files/301671/results.txt)

**Comment by pwittrock** [comment]
@timstclair Need to look into the suite being rerun.  Do the tests pass after you wait a little bit?  Is it possible the etcd port just hasn't been made available yet?

**Comment by pwittrock** [comment]
I bet I know what the issue you encountered is.  I fixed an issue this morning where if images was empty it would default to the default image, so then you want get both.  Looking at the test logs it tries to run the tests against the same host 2x with 1 failure and 1 success.  The error is output twice, once as results and once as the non-0 exit failure message.

File: test/e2e_node/runner/run_e2e.go (hunk 6 above)
Comment (pwittrock) [inline test/e2e_node/runner/run_e2e.go]: Yup.  Good idea.

File: hack/e2e-node-test.sh (hunk 7 above)
Comment (pwittrock) [inline hack/e2e-node-test.sh]: Thanks, I was look for that. :)

**Comment by timstclair** [comment]
Ah, thanks! I'll sync to the latest version and try again.

File: test/e2e_node/runner/run_e2e.go (hunk 10 above)
Comment (pwittrock) [inline test/e2e_node/runner/run_e2e.go]: done

File: test/e2e_node/runner/run_e2e.go (hunk 9 above)
Comment (pwittrock) [inline test/e2e_node/runner/run_e2e.go]: Done

File: test/e2e_node/e2e_build.go (hunk 8 above)
Comment (pwittrock) [inline test/e2e_node/e2e_build.go]: Done

**Comment by pwittrock** [comment]
PTAL

**Comment by pwittrock** [comment]
@k8s-bot node e2e test this issue: #IGNORE core failures

**Comment by pwittrock** [comment]
Rebased.  PTAL.

**Comment by timstclair** [comment]
LGTM

**Comment by pwittrock** [comment]
Thanks @timstclair @vishh  :)

**Comment by pwittrock** [comment]
Adding LGTM after squashing commits

**Pull request merged**